pkg/netlink/notify.go

```diff
@@ -0,0 +1,178 @@
+package netlink
+
+import (
+	"log"
+	"reflect"
+
+	"github.com/opiproject/opi-evpn-bridge/pkg/infradb"
+)
+
+// notifyDBChanges notify the database changes
+func notifyDBChanges() {
+	process[routeKey, *RouteStruct](latestRoutes, routes, ROUTE, RouteOperations)
+	process[nexthopKey, *NexthopStruct](latestNexthop, Nexthops, NEXTHOP, NexthopOperations)
+	process[fdbKey, *FdbEntryStruct](latestFDB, fDB, FDB, FdbOperations)
+	process[l2NexthopKey, *L2NexthopStruct](latestL2Nexthop, l2Nexthops, L2NEXTHOP, L2NexthopOperations)
+}
+
+func process[K comparable, V any](latestDB, oldDB map[K]V, eventType int, ops Operations) {
+	oldgenmap = make(map[interface{}]interface{})
+	latestgenmap = make(map[interface{}]interface{})
+	for k, v := range latestDB {
+		latestgenmap[k] = v
+	}
+	for k, v := range oldDB {
+		oldgenmap[k] = v
+	}
+	event := Event{
+		EventType: eventType,
+		Operation:/*Operate{
+			toAdd:    ops.Add,
+			toUpdate: ops.Update,
+			toDelete: ops.Delete,
+		}*/ops,
+	}
+	notify_changes(latestgenmap, oldgenmap, event)
+}
+
+// nolint
+func notify_changes(new_db map[interface{}]interface{}, old_db map[interface{}]interface{}, event Event) {
+	DB2 := old_db
+	DB1 := new_db
+	/* Checking the Updated entries in the netlink db by comparing the individual keys and their corresponding values in old and new db copies
+	   entries with same keys with different values and send the notification to vendor specific module */
+	for k1, v1 := range DB1 {
+		v2, ok := DB2[k1]
+		if !ok {
+			continue
+		}
+		if !deepCheck(v1, v2) {
+			// To Avoid in-correct update notification due to race condition in which metadata is nil in new entry and crashing in dcgw module
+			if event.EventType == ROUTE || event.EventType == NEXTHOP {
+				var status VrfStatusGetter
+				var ok bool
+				status, ok = v1.(VrfStatusGetter)
+				if !ok {
+					log.Printf("Netlink: Invalid Type")
+					continue
+				}
+				if status.GetVrfOperStatus() == infradb.VrfOperStatusToBeDeleted {
+					notifyAddDel(status, event.Operation.Delete)
+					delete(new_db, k1)
+					delete(old_db, k1)
+					continue
+				}
+			}
+			notifyAddDel(v1, event.Operation.Update)
+		}
+		delete(new_db, k1)
+		delete(old_db, k1)
+	}
+	for _, r := range new_db { // Added entries notification cases
+		notifyAddDel(r, event.Operation.Add)
+	}
+	for _, r := range old_db { // Deleted entires notification cases
+		notifyAddDel(r, event.Operation.Delete)
+	}
+}
+
+func deepCheck(v1 interface{}, v2 interface{}) bool {
+	if reflect.TypeOf(v1) != reflect.TypeOf(v2) {
+		return true
+	}
+	switch t := v1.(type) {
+	case *RouteStruct:
+		return t.deepEqual(v2.(*RouteStruct), true)
+	case *NexthopStruct:
+		return t.deepEqual(v2.(*NexthopStruct), true)
+	case *FdbEntryStruct:
+		return t.deepEqual(v2.(*FdbEntryStruct), true)
+	case *L2NexthopStruct:
+		return t.deepEqual(v2.(*L2NexthopStruct), true)
+	default:
+		return true
```

Comment:
yes, we deal with only the 4 structures defined.

Reply:
and if you add a new type, it will be always true if by mistake this code is not updated?

Reply:
> when a new type is declared, one need to implement the difference check function for it, best thing is we can log in default

best we can do here is to log an error and report that they are not equal if an unknown types are coming

Reply:
Done

Reply:
true is still returned

Reply:
yes, we log and return true because don't want to handle such cases, 
In case the values are equal we return true to deepcheck and we don't process it simply delete it,
same when we get unknown structs we log and return true and delete them